catalog/app/utils/PackageUri.js

```diff
@@ -0,0 +1,90 @@
+import { parse as parseQs } from 'querystring'
+import { parse as parseUrl } from 'url'
+
+import * as R from 'ramda'
+
+import { BaseError } from 'utils/error'
+
+export class PackageUriError extends BaseError {
+  // eslint-disable-next-line react/static-property-placement
+  static displayName = 'PackageUriError'
+
+  constructor(msg, uri) {
+    super(`Invalid package URI (${uri}): ${msg}`, { msg, uri })
+  }
+}
+
+function parsePackageSpec(spec, uri) {
```

Comment:
Can you make returning value consistent? I mean, return value with the same number of properties. If some property is omitted, make it `null`:

```
  if (spec.includes(':')) {
    return { hash: null, name, tag }
  }
  if (spec.includes('@')) {
    return { hash, name, tag: null }
  }
  return { hash: null, name: spec, tag: null }
```

Reply:
but why? missing prop access returns undefined, which seems fine

Reply:
To make code more predictable. I need to read the whole function now to know what are possible values (if I'm interested just in form of data). And if I miss some return, I can get unexpected results.